Topic/add boost primitives

mtmccrea opened the pull request:
This pull request exposes 100+ Special Functions from the Boost library. Documented with both a Help File and a Guide. More details to follow in the thread...

mtmccrea commented:
There's a number of things to consider when reviewing this pull request:

1. I added these functions as methods to `Number`. Would there be a more appropriate class to add these to? Perhaps it's own class? `SimpleNumber`? `SpecialFunction : SimpleNumber`? `SpecialFunction : AbstractFunction`?
2. In the course of exposing the boost primitives, a number of function templates were used. This could likely use some future revision by someone who is more familiar with the C++ primitives, the boost library in particular.  The current implementation though does manage to expose 100+ boost functions with just 5 or 6 templates. For example:
https://github.com/mtmccrea/supercollider/blob/29f30cbd8a22dfeba1239f6c422895b98ce7c49a/lang/LangPrimSource/PyrMathPrim.cpp#L1249
3. I tried to make the SC implementations align as closely to the boost library as possible. This includes:
  - Defering to the Boost docs as the primary source of detailed information. I've documented the basic purpose of each function and argument bounds when useful.
  - Method name and primitive name matches as closely to the boost implementation as possible, despite inconsistencies in the boost library.
  - I let let invalid values be caught by boost (Error is thrown from primitive).
  - - In 3 functions, Boost won't catch an argument that is out of range value and interpreter hangs. I catch this manually in SC. Is this an appropriate way to catch the error? See legendreQ, laguerreAssoc, hermite https://github.com/mtmccrea/supercollider/blob/543683ba96461cd8b5d3c5eba684d41ec492a133/SCClassLibrary/Common/Math/Number.sc#L185
4. The methods are documented in the style of generic functions, e.g. `foo(a, b)` rather than `a.foo(b)`, which makes their correspondence to the underlying boost function more clear. As a result, all of the functions had to be listed as `PRIVATE::` in the help file so they don't show up as "undocumented instance methods".
5. One method added is `Number : factorial`. This does a factorial calculation in floating point, so it raises the resolution hugely over `Integer : factorial`, to where it is actually useful. We might consider deprecating `Integer : factorial`?
6. There is a Guide! Tour of Special Functions. It illustrates the behavior of nearly all of the functions, which I think will be helpful for people to get an idea of how these might be use to ... make sound (But there aren't sound examples :-/).

mtmccrea commented:
**OT**: in the course of generating number series for the examples, I came across a frustrating loss of precision doing things like: 
```
(-2, -1.95 .. 1)
-> [ -2, -1.95, -1.9, -1.85, -1.8, -1.75, -1.7, -1.65, -1.6, -1.55, -1.5, -1.45, -1.4, -1.35, -1.3, -1.25, -1.2, -1.15, -1.1, -1.05, -1, -0.95, -0.9, -0.85, -0.8, -0.75, -0.7, -0.65, -0.6, -0.55, -0.5, -0.45, -0.4, -0.35, -0.3, -0.25, -0.2, -0.15, -0.099999999999998, -0.049999999999998, 1.7763568394003e-15, 0.050000000000002, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1 ]
```
This was a problem when trying to, for example, land exactly on 0 in the series. I know this is to be expected in some situations, but I was a bit surprised that a couple dozen values generated in increments of 0.5 lose precision so quickly, and it's restored on the other side of 0. Just wanted to check to see if anyone sees this as a problem, or totally expected.

joslloand commented:
Comments on @brianlheim thoughts:

> I'd definitely like to see this go into 3.10

_Excellent_ to hear this. The added Boost Special Functions give much of the math we need for HOA... which is [coming along](https://github.com/ambisonictoolkit/atk-sc3/tree/sandbox/hoa).

mtmccrea commented:
Thanks for your thoughts, @brianlheim.

> The appveyor 32-bit build is failing, it looks like you didn't make setSlotVal available to the 32-bit build. I think you can probably just add it to the generic PyrSlot.h header.

The most recent commit adds setSlotVal to PyrSlot32.h. After trying to put it in PyrSlot.h, I ran into a roadblock trying to define the function template there. I looked at the other setters in PyrSlot64.h and saw they're duplicated in PyrSlot32, so I followed this approach and now the appveyor build passes. 

Oddly, now the travis build has failed 2/3 tests and I can't readily spot why. You'll see that this passed [before](https://travis-ci.org/supercollider/supercollider/builds/365922989). I'm suspicious it doesn't relate to this change...ideas?

>This will conflict with the work I've done in #3582.

Good to know. You'll certainly have deeper insights than me on how this PR might integrate with your changes, which seem like a larger revision of the primitive writing mechanism. This will be a big improvement!

>adding even one or two sound examples might be really helpful for people to make the connection and start somewhere. @telephon might be able to help, i believe he enjoys writing sound examples :)

I'll keep this in the backlog, though likely won't get around to adding it to this PR, as this has already become a monumentally larger effort than my original intention of just unlocking boost's Legendre Polynomial ;) In lieu of sound examples, the pretty plots will have to do for the moment...

Again, please let me know if you are able to see if it was this last commit that broke the travis build or...something else.

brianlheim commented:
> I'm suspicious it doesn't relate to this change...ideas?

It doesn't, I just messaged @scztt about this and it's something to do with a breaking "patch" update in one of the Ruby packages used in our test runner. Should be fixed soon. Sorry about that. MSVC 2013 is the most picky compiler we have in CI, and the only one that builds for a 32-bit architecture, so as long as Appveyor passes you're probably fine.

> After trying to put it in PyrSlot.h, I ran into a roadblock trying to define the function template there.

What specifically is the issue? It shouldn't be a problem AFAICT.

> In lieu of sound examples, the pretty plots will have to do for the moment...

No worries, that's fine by me.

mtmccrea commented:
> So the loss of precision where the repeating binary fraction for 0.05 is truncated moves at least 5 bits left, which now is enough to be visible at low magnitudes.

Thanks for the explanation, @jamshark70.

> people will start complaining about seeing inexact numbers that used to print without lengthy decimals... which, here, is exactly what happened.

Didn't mean to come off as complaining... in fact I prefer that precision in _posting_, helps to locate why the imprecise _result_ is breaking a bounded function :) An issue for a different thread...

mtmccrea commented:
@brianlheim 
> > After trying to put it in PyrSlot.h, I ran into a roadblock trying to define the function template there.

> What specifically is the issue? It shouldn't be a problem AFAICT.

Ah, I misunderstood the first time around. I tripped down a rabbit hole trying to set up extern templates—this is outside of my wheelhouse—but moving both the declaration _and_ the definitions into PyrSlot.h seems to have done the trick. Does this look OK?

brianlheim commented:
> but moving both the declaration and the definitions into PyrSlot.h seems to have done the trick. Does this look OK?

Yes, it's pretty common to put template definitions in headers. Thanks!

telephon commented on HelpSource/Guides/Tour-of-Special-Functions.schelp:
```diff
@@ -0,0 +1,1073 @@
+		txt = txt ++ params.asString;
+		params.collect({ |param, i|
+			samps.collect({ |smp|
+				if (func.isKindOf(Symbol)) {
```
you can replace this code by: `func.applyTo(smp, i)`

mtmccrea replied:
Useful! Done.

telephon commented on SCClassLibrary/Common/Math/Number.sc:
```diff
@@ -61,4 +61,326 @@ Number : Magnitude {
+
+	/*	Factorials and Binomial Coefficients */
+
+	// TODO: replace Integer:factorial?
```
yes, if your primitive treats both  `Float` and `Integer` (in sclang) you just remove the current implementation from `Integer.sc`.

mtmccrea replied:
I've removed Integer:factorial and cleared it from the Integer help doc.

mtmccrea commented:
@telephon 
> normally, we expect math that is implemented for numbers to be also implemented for `Array`, `AbstractFunction` and for `UGen`. 

Great suggestion, I was meaning to look into this... I did some digging and I was able to get part of the way there. 
1. Extending it to arrays looks like:
```
jacobiCd { arg k, u; ^this.performBinaryOp('jacobiCd', k, u) }
```
...and that works. Though it looks like other math functions are in `SequenceableCollection` as opposed to `Array`, correct?


2. Adding the functions to `UGen` has me confused. To `AbstractFunction` I added:
```
jacobiCd { arg k, u; ^this.composeBinaryOp('jacobiCd', k, u) }
```
and I got the error: `Operator 'jacobiCd' applied to a UGen is not supported in scsynth`

So three questions:

- Is there something wrong with the syntax?
- Can I expect that adding this to `AbstractFunction` will make it available to `UGen`, as it is a subclass of `AbstractFunction`? In your comment you listed them independently, suggesting it needs to be added to both classes.
- It's notable that other operations that are available to UGens, like `rrand`, which belong to `AbstractFunction` share the same method with `SimpleNumber` instead of `Number`. However moving the method from:
``` Number {
jacobiCd { |u| _JacobiCd; ^this.primitiveFailed }
}
```
to 
``` SimpleNumber {
jacobiCd { arg k, u; _JacobiCd; ^k.performBinaryOpOnSimpleNumber('jacobiCd', this, u) }
}
```
... still gives the same error. Aside from it not yet working with UGen, should I move these all to `SimpleNumber`? (What's the difference?)
Tips welcome!

telephon commented:
>1. 
>Should I add them to Array or SequenceableCollection?

if they are operations that don't operate on a whole array, but on each element one after the other, then `SequenceableCollection`.

>2.
>and I got the error: Operator 'jacobiCd' applied to a UGen is not supported in scsynth

You should just add the methods to `AbstractFunction` and then you need to implement the backend in the primitive file in the right places. I'll check tomorrow where these are. You could try and find them yourself in the meantime tracing a method like `lcm` (this should make a good model).

mtmccrea commented on HelpSource/Classes/Number.schelp:
```diff
@@ -91,3 +92,615 @@ argument:: function
+::
+
+WARNING::
+Many of the functions are only valid in certain numerical ranges. For the most part, error handling
```
Done!

mtmccrea commented:
Thanks for the leads on this, @telephon.

I've gone ahead and extended these methods to `SequenceableCollection`, it's already a great improvement. Note I did come across some issues when trying to document it (#3694). Not a show stopper though.

I looked into `lcm` and `gcd` for extending these methods to `AbstractFunction`. While tracing this out looks simple enough for a method or three, just the sheer number of methods here (100+) means another big push to make this happen, if the methods are to be fully implemented for UGens (aa, kk, ak, ka, etc...). Additionally many of these functions are not defined for certain numerical ranges, so we would need to see how the errors are handled by scsynth when, for example, you cross an asymptote of one of these functions.  It looks like I should be able to do this eventually, but I'm wondering if this could be left to another PR?  

Otherwise I think I've addressed all the issues that have come up in response to this PR so far...

telephon commented:
> It looks like I should be able to do this eventually, but I'm wondering if this could be left to another PR?

Yes, I think it is reasonable to do this in an extra step. 

>Additionally many of these functions are not defined for certain numerical ranges, so we would need to see how the errors are handled by scsynth when, for example, you cross an asymptote of one of these functions.

Indeed, many of the existing functions introduce a special interpretation as to make operators meaningful for signals. It might be a subtle issue how to find such interpretations for each of the new ones. The best way to proceed is probably:

1.  find those which have no restrictions
2. incrementally add interpretations to those which have restrictions (only if the interpretations are reasonable).

I'd be happy in helping in such issues (for `gcd` and `lcm`, this was quite adventurous, but also very enjoyable).

telephon commented on HelpSource/Classes/SimpleNumber.schelp:
```diff
@@ -990,3 +990,621 @@ subsection:: misc
+Below you'll find descriptions of the functions and their bounds, but for
+visualizing the functions, have a look in link::Guides/Tour-of-Special-Functions::.
+
+note::The following methods are documented in the style of a generic function call
```
This is usually called "Functional notation" and "Receiver notation" respectively, and are described in the `Messages` help file. I'm not sure if you need this discussion here really, because everyone is (or should become) acquainted with this style variation. Were we to add it everywhere, it would be a lot of extra text in many helpfiles.

mtmccrea replied:
Good point! 110ca69

mtmccrea replied:
I updated the note in `SequenceableCollection.schelp` as well, though that can be removed down the line once #3694 is addressed.

telephon commented on HelpSource/Guides/Tour-of-Special-Functions.schelp:
```diff
@@ -0,0 +1,1065 @@
+	var res, params, samps, pltr;
+	var gridcol = Color.gray.alpha_(0.3);
+
+	res = if (data.shape.size == 1) { // 1D data
```
I'm pretty sure you could simplify this, by factoring out the two identical inner parts. But I'm fine with it as it is now if you don't have the time for it.

mtmccrea replied:
The inner parts weren't quite identical, but I was able to refactor it. 7b2be56

telephon commented on HelpSource/Guides/Tour-of-Special-Functions.schelp:
```diff
@@ -0,0 +1,1065 @@
+var plotSpecs = [
+	[   // tgamma(z)
+		{|z| // make sure z isn't a negative integer
+			if (((z<=0) and: ((z % 1) ==0)).not) { tgamma(z) }
```
for efficiency, this should be:
` 		if (((z<=0) and: { ((z % 1) ==0)).not }) { tgamma(z) } `
(it will inline, same the other cases below)

and in this case of course:

` 		if (((z<=0) and: { (z % 1) !=0) }) { tgamma(z) } `

telephon replied:
even though it won't inline as nicely, I'd also tend to make a function:

```supercollider
var call = { |x, selector|  // make sure x isn't a negative integer
if((x <=0) and: { (x % 1) !=0) }) { x.perform(selector) }
};
```

and then just call this function further down, e.g.

```supercollider
{|dz| call.value(dz, \tgamma1pm1) }
```

telephon commented:
>Help is certainly appreciated, so feel free if you'd like to!

When you want to start working on the next step, just open a PR and I'll join for discussion.

>So I believe this addresses all the requested changes for this PR thus far.

yes, more than that! Thank you.

telephon commented:
I've just noticed a few issues that have come up with this. One is a behavior change in `factorial`, which returns a float now also when called on an integer. I'll add a fix for this.

There are some other things which had escaped me, I'll make some issues, I hope you can fix them easily.

joslloand commented:
@telephon wrote:

> I've just noticed a few issues that have come up with this. One is a behavior change in factorial, which returns a float now also when called on an integer. I'll add a fix for this.

FYI, one of the issues affecting our HOA implementation that we are seeing with Integer factorial is overflow. That is, if the returned result on an integer was outside the precision of integer, we'd receive an incorrect value. (@mtmccrea could remind me of whether an error is received, or just an incorrect result.)

Perhaps (?) an ideal behavior would be to return a float when integer would overflow?

telephon commented:
The recommended way to deal with such precision issues is to use a float instead. Do you think you can fix it on your side by just replacing the receiver with a float? The reply will also be a float, 64-bit.

joslloand commented:
Yes, that should be possible as a cast to float where need be. Thanks.
